### DISCUSSION TITLE
Build Kernels concurrently using the same KernelBuilder

### DISCUSSION THREAD

**Discussion by songpeidu** [microsoft/semantic-kernel#5372] [Q&A] [answered]
It looks like `IKernelBuilder.Build` is not thread safe, is this by design?

**Comment by songpeidu** [comment]
Saw the comments regarding `GetKeyedServices`, so after that's fixed, it should be thread safe right?

**Comment by madsbolaris** [comment, marked as answer]
Yes, this by design. You ideally shouldn't be using a builder to create multiple Kernels. Ideally, you would either...
- Create a kernel once and then clone it.
- Add a transient kernel to the service provider so that it's always recreated.

For the latter, I'd refer to the Dependency Injection sample: https://github.com/microsoft/semantic-kernel/tree/main/dotnet/samples/HomeAutomation

**Reply by thomasdc** [reply]
This limitation is the potential root cause of our failing unit tests. In those tests we do use several KernelMemoryBuilders. Strangely enough marking the assembly as NonParallelizable does not fix the issue.
Any recommendations for unit test scenario's?
